### PR TITLE
Throw an exception if &make is called w/o a valid match in $/

### PR DESCRIPTION
**Pull request by Kaiepi**
This adds a new `X::MakeWithInvalidMatch` exception, which is thrown by
`&make` when `$/` is not an `NQPMatchRole`. This fixes a LTA error when
`&make` is called, but `$/` hasn't been given a value.

Passes `make test` and `make spectest` (outside of https://github.com/rakudo/rakudo/issues/4027).

### CODE REVIEW

File: src/core.c/Match.pm6 [hunk 1]
```diff
@@ -322,10 +332,11 @@ multi sub infix:<eqv>(Match:D \a, Match:D \b) {
     );
 }
 
-
 sub make(Mu \made) {
-    nqp::bindattr(nqp::decont(nqp::getlexcaller('$/')),Match,'$!made',made)
+    nqp::if(
+      nqp::istype((my Mu:_ $match := nqp::decont(nqp::getlexcaller('$/'))),NQPMatchRole),
+      nqp::bindattr($match,Match,'$!made',made),
```
Comment (Kaiepi) [inline src/core.c/Match.pm6:338]: This seems to work with slangs... but how?

**Comment by lizmat** [comment]
Do you have any idea on performance issues of this change?  `make` is called an awful lot when e.g. compiling the setting.

**Comment by Kaiepi** [comment]
Compiling the setting takes around the same amount of time with or without this for me, but in a benchmark, `&make` with a `Match` in `$/` runs around 10% slower than it did before. `NQPClassHOW` and `Metamodel::GrammarHOW` would have `NQPMatchRole` in their typecheck cache, so I think this checks out.

**Comment by lizmat** [comment]
To be honest, I'm a bit reluctant to merge this because:

a. grammars are slow as it is, so making them slower again feels wrong
b. this seems inspired by a flapper causing `$/` to become undefined when it shouldn't

So the real life case for this error to appear to be very limited.  When you're setting `$/` yourself, then you either know what you're doing, or it is a case of DIWHIDT.

So I'm leaning towards -1 for this PR.  @jnthn maybe has other opinions.  If he doesn't merge in the next few days, I will close this PR.

**Pull request closed without merging**